src/Request.ts

```diff
@@ -0,0 +1,320 @@
+		var start = kwArgs.start,
+			end = kwArgs.end,
+			requestArgs: dstore.FetchArgs = {};
+		if (this.useRangeHeaders) {
+			requestArgs.headers = lang.mixin(this._renderRangeHeaders(start, end), kwArgs.headers);
+		} else {
+			requestArgs.queryParams = this._renderRangeParams(start, end);
+			if (kwArgs.headers) {
+				requestArgs.headers = kwArgs.headers;
+			}
+		}
+
+		var results = this._request(requestArgs);
+		return <dstore.FetchPromise<T>>QueryResults(results.data, {
+			totalLength: results.total,
+			response: results.response
+		});
+	}
+
+	_request(kwArgs: dstore.FetchArgs) {
+		kwArgs = kwArgs || {};
+
+		// perform the actual query
+		var headers: { [ name: string ]: string } = <{ [ name: string ]: string }>lang.mixin(Object.create(this.headers), { Accept: this.accepts });
+
+		if ('headers' in kwArgs) {
+			lang.mixin(headers, kwArgs.headers);
+		}
+
+		var requestUrl = this._renderUrl(kwArgs.queryParams);
+
+		var response = request(requestUrl, {
+			method: 'GET',
+			headers: headers
+		});
+		var collection = this;
+		var parsedResponse = response.then(function (response: Response<string>) {
+			return collection.parse(response.data);
+		});
+		return {
+			data: parsedResponse.then(function (data) {
+				// support items in the results
+				var results = data.items || data;
+				for (var i = 0, l = results.length; i < l; i++) {
+					results[i] = collection._restore(results[i], true);
+				}
+				return results;
+			}),
+			total: parsedResponse.then(function (data) {
+				// check for a total property
+				var total = data.total;
+				if (total > -1) {
+					// if we have a valid positive number from the data,
+					// we can use that
+					return total;
+				}
+				// else use headers
+				return response.then(function (response) {
+					var rangeHeader = response.getHeader('Content-Range');
+					let rangeMatch: RegExpMatchArray;
+					let range: number;
+					if (rangeHeader && (rangeMatch = rangeHeader.match(/\/(.*)/)) && (range = Number(rangeMatch[1]))) {
+						return range;
+					} else {
+						return null;
+					}
+				});
+			}),
+			response: response
+		};
+	}
+
+	_renderFilterParams(filter: Filter) {
+		// summary:
+		//		Constructs filter-related params to be inserted into the query string
+		// returns: String
+		//		Filter-related params to be inserted in the query string
+		var type = filter.type;
+		var args = filter.args;
+		if (!type) {
+			return [''];
+		}
+		if (type === 'string') {
+			return [args[0]];
+		}
+		if (type === 'and' || type === 'or') {
+			return [filter.args.map(function (arg) {
+				// render each of the arguments to and or or, then combine by the right operator
+				var renderedArg = this._renderFilterParams(arg);
+				return ((arg.type === 'and' || arg.type === 'or') && arg.type !== type) ?
+					// need to observe precedence in the case of changing combination operators
+					'(' + renderedArg + ')' : renderedArg;
+			}, this).join(type === 'and' ? '&' : '|')];
```

Comment:
this whole block is kind of hard to read. One simple thing that might help it a little is to set the join token in a variable. So instead of `join(type === 'and' ? '&' : '|')` you could just do `join(joinToken)` and assign `joinToken` up above.